File: leetcode/easy/108/108.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func build(nums []int, lo, hi int) *TreeNode {
	if lo > hi {
		return nil
	}
	mid := (lo + hi) / 2
	return &TreeNode{
		Val:   nums[mid],
		Left:  build(nums, lo, mid-1),
		Right: build(nums, mid+1, hi),
	}
}

func sortedArrayToBST(nums []int) *TreeNode {
	return build(nums, 0, len(nums)-1)
}

File: leetcode/easy/108/108.rs
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub struct Solution;

impl Solution {
    pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        fn build(nums: &[i32], lo: i32, hi: i32) -> Option<Rc<RefCell<TreeNode>>> {
            if lo > hi {
                return None;
            }
            let mid = ((lo + hi) / 2) as usize;
            let mut node = TreeNode::new(nums[mid]);
            node.left = build(nums, lo, mid as i32 - 1);
            node.right = build(nums, mid as i32 + 1, hi);
            Some(Rc::new(RefCell::new(node)))
        }
        build(&nums, 0, nums.len() as i32 - 1)
    }
}
